src/grammar.rs

```diff
@@ -5,11 +5,108 @@ use crate::production::Production;
 use crate::term::Term;
 use rand::{rngs::StdRng, seq::SliceRandom, thread_rng, Rng, SeedableRng};
 use serde::{Deserialize, Serialize};
-use stacker;
+
 use std::fmt;
 use std::slice;
 use std::str;
 
+/// A node of a `ParseTree`, either terminating or continuing the `ParseTree`
+#[derive(Debug, Clone)]
+pub enum ParseTreeNode<'gram> {
+    Terminal(&'gram str),
+    Nonterminal(ParseTree<'gram>),
+}
+
+#[derive(Debug, Clone)]
+pub struct ParseTree<'gram> {
+    pub lhs: &'gram Term,
+    pub rhs: Vec<ParseTreeNode<'gram>>,
```

Comment:
IMO this shouldn't be public, but should be exposed via iterators like the other BNF types. what do you think @shnewto ?